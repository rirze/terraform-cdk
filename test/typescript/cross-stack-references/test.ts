import * as fs from "fs-extra";
import * as path from "path";
import { TestDriver } from "../../test-helper";

describe("cross stack references", () => {
  let driver: TestDriver;

  beforeAll(async () => {
    driver = new TestDriver(__dirname);
    await driver.setupTypescriptProject();
    await driver.synth();
  });

  test("synth generates JSON", () => {
    expect(driver.manifest()).toMatchSnapshot();
  });

  describe("deployed", () => {
    beforeAll(async () => {
      await driver.deploy("source");
      await driver.deploy("passthrough");
      await driver.deploy("sink");
    });

    it("references primitive values", () => {
      expect(driver.readLocalFile("originNum")).toBe(
        driver.readLocalFile("passthroughNum")
      );
      expect(driver.readLocalFile("originStr")).toBe(
        driver.readLocalFile("passthroughStr")
      );
    });

    it("references can be passed through stacks", () => {
      expect(driver.readLocalFile("originNum")).toBe(
        driver.readLocalFile("sinkNum")
      );
      expect(driver.readLocalFile("originStr")).toBe(
        driver.readLocalFile("sinkStr")
      );
    });

    it("references complex values", () => {
      expect(driver.readLocalFile("originList")).toBe(
        driver.readLocalFile("passthroughList")
      );
    });

    describe("terraform functions", () => {
      beforeAll(async () => {
        await driver.deploy("fns");
        await driver.deploy("functionOutput");
      });

      it("can use reference in terraform function", () => {
        // num: Fn.sum([src.num, src.num]),
        const originNum = parseInt(driver.readLocalFile("originNum"), 10);
        const result = parseInt(driver.readLocalFile("fnsNum"), 10);

        expect(originNum * 2).toBe(result);
      });

      it("can use references from two stacks", () => {
        // str: Fn.join(",", [src.str, passthrough.str]),
        const originStr = driver.readLocalFile("originStr");
        const passthroughStr = driver.readLocalFile("passthroughStr");

        const result = driver.readLocalFile("fnsStr");
        expect(`${originStr},${passthroughStr}`).toBe(result);
      });

      it("references terraform function output", () => {
        const originNum = parseInt(driver.readLocalFile("originNum"), 10);
        const result = parseInt(driver.readLocalFile("functionOutputNum"), 10);

        expect(originNum * 2).toBe(result);
      });
    });
  });
});
