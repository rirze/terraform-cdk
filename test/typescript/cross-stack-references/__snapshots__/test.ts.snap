// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cross stack references synth generates JSON 1`] = `
"{
  \\"version\\": \\"stubbed\\",
  \\"stacks\\": {
    \\"source\\": {
      \\"name\\": \\"source\\",
      \\"constructPath\\": \\"source\\",
      \\"workingDirectory\\": \\"stacks/source\\",
      \\"synthesizedStackPath\\": \\"stacks/source/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    },
    \\"passthrough\\": {
      \\"name\\": \\"passthrough\\",
      \\"constructPath\\": \\"passthrough\\",
      \\"workingDirectory\\": \\"stacks/passthrough\\",
      \\"synthesizedStackPath\\": \\"stacks/passthrough/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"source\\"
      ]
    },
    \\"sink\\": {
      \\"name\\": \\"sink\\",
      \\"constructPath\\": \\"sink\\",
      \\"workingDirectory\\": \\"stacks/sink\\",
      \\"synthesizedStackPath\\": \\"stacks/sink/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"source\\"
      ]
    },
    \\"fns\\": {
      \\"name\\": \\"fns\\",
      \\"constructPath\\": \\"fns\\",
      \\"workingDirectory\\": \\"stacks/fns\\",
      \\"synthesizedStackPath\\": \\"stacks/fns/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"source\\"
      ]
    },
    \\"functionOutput\\": {
      \\"name\\": \\"functionOutput\\",
      \\"constructPath\\": \\"functionOutput\\",
      \\"workingDirectory\\": \\"stacks/functionOutput\\",
      \\"synthesizedStackPath\\": \\"stacks/functionOutput/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"source\\"
      ]
    }
  }
}"
`;
